lib/sla.js

```diff
@@ -45,80 +48,135 @@ function designateMixingBridge(client, bridgeName) {
+  // still inside of the shared extension waiting for others to answer.
+  if (bridge.channels.length === 1) {
+    bridge.stopMoh();
+    updateState(client, bridge.name, inUse);
+  }
+
+  // Resets listeners, so that redundant events aren't accepted.
+  bridge.removeAllListeners('ChannelEnteredBridge');
+  bridge.removeAllListeners('ChannelLeftBridge');
   var defer = Q.defer();
-  var numInBridge = 0;
 
   console.log('Adding channels to bridge %s', bridge.id);
 
   var addChannel = Q.denodeify(bridge.addChannel.bind(bridge));
-  addChannel({channel: [inbound.id, dialed.id]})
+  addChannel({channel: channel.id})
     .catch(function (err) {
       defer.reject(err);
     });
 
+  if (dialed) {
+    addChannel({channel: dialed.id})
+      .catch(function (err) {
+        defer.reject(err);
+      });
+  }
+  bridge.on('ChannelEnteredBridge', channelEntered);
+  bridge.on('ChannelLeftBridge', channelLeft);
+  return defer.promise;
+
+  /**
+   * Function that displays a helpful message when a channel enters a bridge
+   * @param {Object} event - the event related to this bridge entering
+   * @param {Object} object - contains the bridge and channel in question
+   */
   function channelEntered(event, object) {
+    object.channel.on('ChannelHold', channelHold);
     console.log('Channel %s has entered the bridge', object.channel.id);
-    numInBridge += 1;
-  }
 
-  bridge.on('ChannelEnteredBridge', channelEntered);
+  }
 
+  /**
+   * Function that is utilized when a channel leaves the bridge.
+   *   If the channel is a non-station channel, then hangs up remaining
+   *   channels.  If the channel is a station, checks if the only non-station
+   *   channel is the only left, and then hangs up that non-station channel.
+   * @param {Object} event - the event related to the bridge leaving
+   * @param {Object} object - contains the bridge and channel in question
+   */
   function channelLeft(event, object) {
-    numInBridge -= 1;
-    console.log('Channel %s has left the bridge', object.channel.id);
-
-    if (numInBridge === 1) {
-      if (object.channel.id === inbound.id) {
-        console.log('Hanging up dialed channel %s', dialed.id);
-        var hangupDialed = Q.denodeify(dialed.hangup.bind(dialed));
-        hangupDialed()
-          .catch(function (err) {
-            err.name = 'HangupFailure';
-            defer.reject(err);
-          });
-      }
-      else if (object.channel.id === dialed.id) {
-        console.log('Hanging up inbound channel %s', inbound.id);
-        var hangupInbound = Q.denodeify(inbound.hangup.bind(inbound));
-        hangupInbound()
-          .catch(function (err) {
-            err.name = 'HangupFailure';
-            defer.reject(err);
+    object.channel.removeAllListeners('ChannelHold');
+
+    isStation(extension, object.channel).then(function (isStation) {
+      if (isStation) {
+        if (object.bridge.channels.length === 1) {
```

Comment:
Two issues with this:

1) You aren't checking to see if the last channel in the bridge is a station. Right now, this will start playing MoH to another station, which isn't correct.
2) You aren't checking that a station left because it pressed hold. It looks like this routine is called any time a channel leaves a bridge. They can leave a bridge because they hung up, in which case, you shouldn't play MoH to the trunks - they aren't on hold, they were hung up on. In that case, they should probably be ejected from the shared extension.